File: android/src/main/kotlin/com/amwalpay/ecr/flutter/EcrTerminalPort.kt
package com.amwalpay.ecr.flutter

import com.amwalpay.ecr.EcrInquiry
import com.amwalpay.ecr.EcrReachability
import com.amwalpay.ecr.EcrReceipt
import com.amwalpay.ecr.EcrResult
import java.math.BigDecimal

/**
 * The part of the ECR SDK this wrapper uses.
 *
 * `EcrTerminal` and `EcrWebServiceTerminal` are final classes, so a seam is
 * needed to test the handler without a terminal on the network.
 */
internal interface EcrTerminalPort {

    suspend fun isReachable(): Boolean

    suspend fun probeReachability(): EcrReachability

    suspend fun sale(amount: BigDecimal, merchantReference: String): EcrResult

    suspend fun void(
        receiptNumber: String,
        originalTerminalId: String,
        merchantReference: String,
    ): EcrResult

    suspend fun refund(
        amount: BigDecimal,
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String,
    ): EcrResult

    suspend fun inquire(
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String,
    ): EcrInquiry

    suspend fun inquireByReference(
        originalReference: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String,
    ): EcrInquiry

    suspend fun receipt(
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String,
    ): EcrReceipt
}

/** Builds the port for one call's terminal. Replaced in tests. */
internal fun interface EcrTerminalFactory {
    fun create(
        host: String,
        serialNumber: String,
        transport: String,
        config: com.amwalpay.ecr.EcrConfig,
    ): EcrTerminalPort
}

File: ios/amwal_ecr/Sources/amwal_ecr/EcrTerminalPort.swift
import Foundation
import AmwalECR

/// The part of an ECR implementation this wrapper uses.
///
/// The seam. Satisfied by [SdkLanTerminal], [SdkWebServiceTerminal], and test
/// doubles. Swapping in a different implementation means conforming a type to
/// this protocol and changing [EcrTerminalFactory]; the handler, the mapping,
/// the channel contract and the whole Dart API stay exactly as they are.
protocol EcrTerminalPort: AnyObject {

    func isReachable() -> Bool

    func sale(amount: Decimal, merchantReference: String) throws -> EcrResult

    func void(
        receiptNumber: String,
        originalTerminalId: String,
        merchantReference: String
    ) throws -> EcrResult

    func refund(
        amount: Decimal,
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String
    ) throws -> EcrResult

    func inquire(
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String
    ) throws -> EcrInquiry

    func inquireByReference(
        _ originalReference: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String
    ) throws -> EcrInquiry

    func receipt(
        receiptNumber: String,
        transactionDate: String,
        originalTerminalId: String,
        merchantReference: String
    ) throws -> EcrReceipt

    func cancel()
}

/// Builds the port for one call's terminal. Replaced in tests.
typealias EcrTerminalFactory = (
    _ host: String,
    _ serialNumber: String,
    _ transport: String,
    _ config: EcrConfig
) -> EcrTerminalPort

func makeNativeTerminal(
    host: String,
    serialNumber: String,
    transport: String,
    config: EcrConfig
) -> EcrTerminalPort {
    EcrSessionPorts.create(
        host: host,
        serialNumber: serialNumber,
        transport: transport,
        config: config
    )
}
